File: hdf-generators/go/merge_test.go
package generators

import (
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
)

func strPtr(s string) *string { return &s }

func TestMergeRequirement_ScalarFieldsTakeUpstream(t *testing.T) {
	current := hdf.BaselineRequirement{
		ID:     "V-001",
		Title:  strPtr("Old title"),
		Impact: 0.5,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "Old description"},
		},
	}
	upstream := hdf.BaselineRequirement{
		ID:     "SV-001",
		Title:  strPtr("New title"),
		Impact: 0.7,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "New description"},
		},
	}

	merged := MergeRequirement(current, upstream, "")

	assert.Equal(t, "SV-001", merged.ID, "ID should always come from upstream")
	assert.Equal(t, "New title", *merged.Title, "title should come from upstream by default")
	assert.Equal(t, 0.7, merged.Impact, "impact should come from upstream by default")
}

func TestMergeRequirement_CodePreservedFromCurrent(t *testing.T) {
	code := "  describe command('test') do\n  end"
	current := hdf.BaselineRequirement{
		ID:     "V-001",
		Impact: 0.5,
		Tags:   map[string]any{},
		Code:   &code,
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}
	upstream := hdf.BaselineRequirement{
		ID:     "SV-001",
		Impact: 0.7,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}

	merged := MergeRequirement(current, upstream, "")

	assert.NotNil(t, merged.Code, "code should be preserved from current")
	assert.Equal(t, code, *merged.Code)
}

func TestMergeRequirement_PreferCurrent_ScalarsFromCurrent(t *testing.T) {
	current := hdf.BaselineRequirement{
		ID:     "V-001",
		Title:  strPtr("Old title"),
		Impact: 0.5,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "Old description"},
		},
	}
	upstream := hdf.BaselineRequirement{
		ID:     "SV-001",
		Title:  strPtr("New title"),
		Impact: 0.7,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "New description"},
		},
	}

	merged := MergeRequirement(current, upstream, "current")

	assert.Equal(t, "SV-001", merged.ID, "ID always comes from upstream")
	assert.Equal(t, "Old title", *merged.Title, "title should come from current with --prefer current")
	assert.Equal(t, 0.5, merged.Impact, "impact should come from current with --prefer current")
}

func TestMergeRequirement_PreferUpstream_CodeFromUpstream(t *testing.T) {
	oldCode := "  # old test"
	newCode := "  # new test"
	current := hdf.BaselineRequirement{
		ID:     "V-001",
		Impact: 0.5,
		Tags:   map[string]any{},
		Code:   &oldCode,
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}
	upstream := hdf.BaselineRequirement{
		ID:     "SV-001",
		Impact: 0.7,
		Tags:   map[string]any{},
		Code:   &newCode,
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}

	merged := MergeRequirement(current, upstream, "upstream")

	assert.NotNil(t, merged.Code)
	assert.Equal(t, newCode, *merged.Code, "code should come from upstream with --prefer upstream")
}

func TestMergeTags_UnionUpstreamWinsDefault(t *testing.T) {
	current := map[string]any{
		"cci":    []any{"CCI-000001"},
		"custom": "my-value",
		"gtitle": "Old SRG title",
	}
	upstream := map[string]any{
		"cci":     []any{"CCI-000001", "CCI-000002"},
		"gtitle":  "New SRG title",
		"stig_id": "RHEL-09-001234",
	}

	merged := MergeTags(current, upstream, "")

	assert.Contains(t, merged, "custom", "current-only keys should be preserved")
	assert.Equal(t, "my-value", merged["custom"])
	assert.Equal(t, "New SRG title", merged["gtitle"], "upstream should win on conflict")
	assert.Contains(t, merged, "stig_id", "upstream-only keys should be added")
	assert.Equal(t, upstream["cci"], merged["cci"], "upstream wins on key conflict")
}

func TestMergeTags_UnionCurrentWinsOnPreferCurrent(t *testing.T) {
	current := map[string]any{
		"cci":    []any{"CCI-000001"},
		"custom": "my-value",
		"gtitle": "Old SRG title",
	}
	upstream := map[string]any{
		"cci":     []any{"CCI-000001", "CCI-000002"},
		"gtitle":  "New SRG title",
		"stig_id": "RHEL-09-001234",
	}

	merged := MergeTags(current, upstream, "current")

	assert.Equal(t, "Old SRG title", merged["gtitle"], "current should win on conflict with --prefer current")
	assert.Equal(t, current["cci"], merged["cci"], "current should win on CCI conflict")
	assert.Contains(t, merged, "stig_id", "upstream-only keys should still be added")
}

func TestMergeTags_PreferUpstreamReplacesAll(t *testing.T) {
	current := map[string]any{
		"custom": "my-value",
		"gtitle": "Old SRG title",
	}
	upstream := map[string]any{
		"gtitle":  "New SRG title",
		"stig_id": "RHEL-09-001234",
	}

	merged := MergeTags(current, upstream, "upstream")

	assert.NotContains(t, merged, "custom", "current-only keys should be dropped with --prefer upstream")
	assert.Equal(t, "New SRG title", merged["gtitle"])
	assert.Equal(t, "RHEL-09-001234", merged["stig_id"])
}

func TestMergeDescriptions_UnionByLabelUpstreamWins(t *testing.T) {
	current := []hdf.Description{
		{Label: "default", Data: "Old default"},
		{Label: "check", Data: "Old check"},
		{Label: "custom", Data: "My custom desc"},
	}
	upstream := []hdf.Description{
		{Label: "default", Data: "New default"},
		{Label: "check", Data: "New check"},
		{Label: "fix", Data: "New fix"},
	}

	merged := MergeDescriptions(current, upstream, "")

	// Should have all unique labels: default, check, custom, fix
	labels := make(map[string]string)
	for _, d := range merged {
		labels[d.Label] = d.Data
	}
	assert.Len(t, labels, 4)
	assert.Equal(t, "New default", labels["default"], "upstream wins on label conflict")
	assert.Equal(t, "New check", labels["check"], "upstream wins on label conflict")
	assert.Equal(t, "My custom desc", labels["custom"], "current-only labels preserved")
	assert.Equal(t, "New fix", labels["fix"], "upstream-only labels added")
}

func TestMergeDescriptions_CurrentWinsOnPreferCurrent(t *testing.T) {
	current := []hdf.Description{
		{Label: "default", Data: "Old default"},
	}
	upstream := []hdf.Description{
		{Label: "default", Data: "New default"},
		{Label: "fix", Data: "New fix"},
	}

	merged := MergeDescriptions(current, upstream, "current")

	labels := make(map[string]string)
	for _, d := range merged {
		labels[d.Label] = d.Data
	}
	assert.Len(t, labels, 2)
	assert.Equal(t, "Old default", labels["default"], "current wins on conflict with --prefer current")
	assert.Equal(t, "New fix", labels["fix"], "upstream-only labels still added")
}

func TestMergeDescriptions_PreferUpstreamReplacesAll(t *testing.T) {
	current := []hdf.Description{
		{Label: "default", Data: "Old default"},
		{Label: "custom", Data: "My custom"},
	}
	upstream := []hdf.Description{
		{Label: "default", Data: "New default"},
		{Label: "fix", Data: "New fix"},
	}

	merged := MergeDescriptions(current, upstream, "upstream")

	labels := make(map[string]string)
	for _, d := range merged {
		labels[d.Label] = d.Data
	}
	assert.Len(t, labels, 2, "should only have upstream labels")
	assert.Equal(t, "New default", labels["default"])
	assert.Equal(t, "New fix", labels["fix"])
	assert.NotContains(t, labels, "custom", "current-only labels dropped with --prefer upstream")
}

func TestMergeRefs_UnionDeduplicated(t *testing.T) {
	url1 := "https://example.com/ref1"
	url2 := "https://example.com/ref2"
	url3 := "https://example.com/ref3"
	current := []hdf.Reference{
		{URL: &url1},
		{URL: &url2},
	}
	upstream := []hdf.Reference{
		{URL: &url2},
		{URL: &url3},
	}

	merged := MergeRefs(current, upstream, "")

	assert.Len(t, merged, 3, "should union and deduplicate by URL")
}

func TestMergeRefs_PreferCurrentKeepsCurrent(t *testing.T) {
	url1 := "https://example.com/ref1"
	url2 := "https://example.com/ref2"
	current := []hdf.Reference{{URL: &url1}}
	upstream := []hdf.Reference{{URL: &url2}}

	merged := MergeRefs(current, upstream, "current")

	assert.Len(t, merged, 1)
	assert.Equal(t, url1, *merged[0].URL)
}

func TestMergeRefs_PreferUpstreamKeepsUpstream(t *testing.T) {
	url1 := "https://example.com/ref1"
	url2 := "https://example.com/ref2"
	current := []hdf.Reference{{URL: &url1}}
	upstream := []hdf.Reference{{URL: &url2}}

	merged := MergeRefs(current, upstream, "upstream")

	assert.Len(t, merged, 1)
	assert.Equal(t, url2, *merged[0].URL)
}

func TestMergeRequirement_SeverityFromUpstream(t *testing.T) {
	high := hdf.SeverityHigh
	medium := hdf.Medium
	current := hdf.BaselineRequirement{
		ID:       "V-001",
		Impact:   0.5,
		Severity: &medium,
		Tags:     map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}
	upstream := hdf.BaselineRequirement{
		ID:       "SV-001",
		Impact:   0.7,
		Severity: &high,
		Tags:     map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}

	merged := MergeRequirement(current, upstream, "")

	assert.NotNil(t, merged.Severity)
	assert.Equal(t, hdf.SeverityHigh, *merged.Severity, "severity should come from upstream by default")
}

func TestMergeRequirement_NilCodeOnBothSides(t *testing.T) {
	current := hdf.BaselineRequirement{
		ID:     "V-001",
		Impact: 0.5,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}
	upstream := hdf.BaselineRequirement{
		ID:     "SV-001",
		Impact: 0.7,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}

	merged := MergeRequirement(current, upstream, "")

	assert.Nil(t, merged.Code, "code should be nil when neither side has code")
}

func TestMergeRequirement_SourceLocationFromUpstream(t *testing.T) {
	line := float64(42)
	ref := "controls/test.rb"
	current := hdf.BaselineRequirement{
		ID:     "V-001",
		Impact: 0.5,
		Tags:   map[string]any{},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}
	upstream := hdf.BaselineRequirement{
		ID:     "SV-001",
		Impact: 0.7,
		Tags:   map[string]any{},
		SourceLocation: &hdf.SourceLocation{
			Line: &line,
			Ref:  &ref,
		},
		Descriptions: []hdf.Description{
			{Label: "default", Data: "desc"},
		},
	}

	merged := MergeRequirement(current, upstream, "")

	assert.NotNil(t, merged.SourceLocation, "sourceLocation should come from upstream")
	assert.Equal(t, float64(42), *merged.SourceLocation.Line)
}

File: hdf-generators/test/merge.test.ts
import { describe, it, expect } from 'vitest';
import { mergeRequirement, mergeTags, mergeDescriptions, mergeRefs } from '../src/merge.js';
import { makeRequirement } from './helpers.js';
import type { Reference } from '@mitre/hdf-schema';

describe('mergeRequirement', () => {
  it('should take scalars from upstream by default', () => {
    const current = makeRequirement({ id: 'V-001', title: 'Old title', impact: 0.5 });
    const upstream = makeRequirement({ id: 'SV-001', title: 'New title', impact: 0.7 });

    const merged = mergeRequirement(current, upstream);

    expect(merged.id).toBe('SV-001');
    expect(merged.title).toBe('New title');
    expect(merged.impact).toBe(0.7);
  });

  it('should preserve code from current by default', () => {
    const current = makeRequirement({
      id: 'V-001',
      code: "  describe command('test') do\n  end",
    });
    const upstream = makeRequirement({ id: 'SV-001' });

    const merged = mergeRequirement(current, upstream);

    expect(merged.code).toBe("  describe command('test') do\n  end");
  });

  it('should take scalars from current with prefer=current', () => {
    const current = makeRequirement({ id: 'V-001', title: 'Old title', impact: 0.5 });
    const upstream = makeRequirement({ id: 'SV-001', title: 'New title', impact: 0.7 });

    const merged = mergeRequirement(current, upstream, 'current');

    expect(merged.id).toBe('SV-001');
    expect(merged.title).toBe('Old title');
    expect(merged.impact).toBe(0.5);
  });

  it('should take code from upstream with prefer=upstream', () => {
    const current = makeRequirement({ id: 'V-001', code: '  # old test' });
    const upstream = makeRequirement({ id: 'SV-001', code: '  # new test' });

    const merged = mergeRequirement(current, upstream, 'upstream');

    expect(merged.code).toBe('  # new test');
  });

  it('should have nil code when neither side has code', () => {
    const current = makeRequirement({ id: 'V-001' });
    const upstream = makeRequirement({ id: 'SV-001' });

    const merged = mergeRequirement(current, upstream);

    expect(merged.code).toBeUndefined();
  });

  it('should take severity from upstream by default', () => {
    const current = makeRequirement({ id: 'V-001', severity: 'medium' });
    const upstream = makeRequirement({ id: 'SV-001', severity: 'high' });

    const merged = mergeRequirement(current, upstream);

    expect(merged.severity).toBe('high');
  });
});

describe('mergeTags', () => {
  it('should union with upstream winning key conflicts by default', () => {
    const current = { cci: ['CCI-000001'], custom: 'my-value', gtitle: 'Old SRG title' };
    const upstream = { cci: ['CCI-000001', 'CCI-000002'], gtitle: 'New SRG title', stig_id: 'RHEL-09-001234' };

    const merged = mergeTags(current, upstream);

    expect(merged.custom).toBe('my-value');
    expect(merged.gtitle).toBe('New SRG title');
    expect(merged.stig_id).toBe('RHEL-09-001234');
    expect(merged.cci).toEqual(['CCI-000001', 'CCI-000002']);
  });

  it('should let current win on conflict with prefer=current', () => {
    const current = { cci: ['CCI-000001'], custom: 'my-value', gtitle: 'Old SRG title' };
    const upstream = { cci: ['CCI-000001', 'CCI-000002'], gtitle: 'New SRG title', stig_id: 'RHEL-09-001234' };

    const merged = mergeTags(current, upstream, 'current');

    expect(merged.gtitle).toBe('Old SRG title');
    expect(merged.cci).toEqual(['CCI-000001']);
    expect(merged.stig_id).toBe('RHEL-09-001234');
  });

  it('should replace all with prefer=upstream', () => {
    const current = { custom: 'my-value', gtitle: 'Old SRG title' };
    const upstream = { gtitle: 'New SRG title', stig_id: 'RHEL-09-001234' };

    const merged = mergeTags(current, upstream, 'upstream');

    expect(merged.custom).toBeUndefined();
    expect(merged.gtitle).toBe('New SRG title');
    expect(merged.stig_id).toBe('RHEL-09-001234');
  });
});

describe('mergeDescriptions', () => {
  it('should union by label with upstream winning on conflict', () => {
    const current = [
      { label: 'default', data: 'Old default' },
      { label: 'check', data: 'Old check' },
      { label: 'custom', data: 'My custom desc' },
    ];
    const upstream = [
      { label: 'default', data: 'New default' },
      { label: 'check', data: 'New check' },
      { label: 'fix', data: 'New fix' },
    ];

    const merged = mergeDescriptions(current, upstream);

    const labels = Object.fromEntries(merged.map(d => [d.label, d.data]));
    expect(Object.keys(labels)).toHaveLength(4);
    expect(labels.default).toBe('New default');
    expect(labels.check).toBe('New check');
    expect(labels.custom).toBe('My custom desc');
    expect(labels.fix).toBe('New fix');
  });

  it('should let current win on conflict with prefer=current', () => {
    const current = [{ label: 'default', data: 'Old default' }];
    const upstream = [
      { label: 'default', data: 'New default' },
      { label: 'fix', data: 'New fix' },
    ];

    const merged = mergeDescriptions(current, upstream, 'current');

    const labels = Object.fromEntries(merged.map(d => [d.label, d.data]));
    expect(Object.keys(labels)).toHaveLength(2);
    expect(labels.default).toBe('Old default');
    expect(labels.fix).toBe('New fix');
  });

  it('should replace all with prefer=upstream', () => {
    const current = [
      { label: 'default', data: 'Old default' },
      { label: 'custom', data: 'My custom' },
    ];
    const upstream = [
      { label: 'default', data: 'New default' },
      { label: 'fix', data: 'New fix' },
    ];

    const merged = mergeDescriptions(current, upstream, 'upstream');

    const labels = Object.fromEntries(merged.map(d => [d.label, d.data]));
    expect(Object.keys(labels)).toHaveLength(2);
    expect(labels.default).toBe('New default');
    expect(labels.fix).toBe('New fix');
    expect(labels.custom).toBeUndefined();
  });
});

describe('mergeRefs', () => {
  it('should union and deduplicate by default', () => {
    const current: Reference[] = [{ url: 'https://example.com/ref1' }, { url: 'https://example.com/ref2' }];
    const upstream: Reference[] = [{ url: 'https://example.com/ref2' }, { url: 'https://example.com/ref3' }];

    const merged = mergeRefs(current, upstream);

    expect(merged).toHaveLength(3);
  });

  it('should keep only current with prefer=current', () => {
    const current: Reference[] = [{ url: 'https://example.com/ref1' }];
    const upstream: Reference[] = [{ url: 'https://example.com/ref2' }];

    const merged = mergeRefs(current, upstream, 'current');

    expect(merged).toHaveLength(1);
    expect(merged[0].url).toBe('https://example.com/ref1');
  });

  it('should keep only upstream with prefer=upstream', () => {
    const current: Reference[] = [{ url: 'https://example.com/ref1' }];
    const upstream: Reference[] = [{ url: 'https://example.com/ref2' }];

    const merged = mergeRefs(current, upstream, 'upstream');

    expect(merged).toHaveLength(1);
    expect(merged[0].url).toBe('https://example.com/ref2');
  });
});
